src/bridge/askpass.c

```diff
@@ -0,0 +1,255 @@
+            {
+              g_message ("couldn't write in askpass: closed connection");
+              return FALSE;
+            }
+          else if (errno != EAGAIN && errno != EINTR)
+            {
+              g_message ("couldn't write in askpass: %d %s", (int)errno, g_strerror (errno));
+              return FALSE;
+            }
+        }
+      else
+        {
+          g_debug ("askpass wrote %d bytes", (gint)res);
+          data += res;
+          len -= res;
+        }
+    }
+
+  return TRUE;
+}
+
+static gboolean
+write_control_message (int fd,
+                       JsonObject *options)
+{
+  gchar *payload;
+  gchar *prefix;
+  gsize length;
+  gboolean ret;
+
+  payload = cockpit_json_write_object (options, &length);
+  prefix = g_strdup_printf ("%" G_GSIZE_FORMAT "\n\n", 1 + length);
+  ret = write_all (fd, prefix, -1) && write_all (fd, payload, length);
+  g_free (prefix);
+  g_free (payload);
+
+  return ret;
+}
+
+int
+main (int argc,
+      char *argv[])
+{
+  GOptionContext *context;
+  JsonObject *request = NULL;
+  JsonObject *reply = NULL;
+  GError *error = NULL;
+  const gchar *env;
+  const gchar *command;
+  const gchar *field;
+  const gchar *response;
+  gchar *user = NULL;
+  gchar *cookie = NULL;
+  gchar *challenge = NULL;
+  gint ret = 1;
+
+  static GOptionEntry entries[] = {
+    { NULL }
+  };
+
+  /* Debugging issues during testing */
+#if WITH_DEBUG
+  signal (SIGABRT, cockpit_test_signal_backtrace);
+  signal (SIGSEGV, cockpit_test_signal_backtrace);
+#endif
+
+  context = g_option_context_new (NULL);
+  g_option_context_add_main_entries (context, entries, NULL);
+  g_option_context_set_description (context, "cockpit-bridge uses cockpit-askpass during password prompts.\n");
+
+  g_option_context_parse (context, &argc, &argv, &error);
+  g_option_context_free (context);
+
+  if (error)
+    {
+      g_printerr ("cockpit-stub: %s\n", error->message);
+      g_error_free (error);
+      return 1;
+    }
+
+  /*
+   * We don't send an init message. This is meant to be used either after an
+   * "init" message has been sent, or with a caller that makes an exception for
+   * the "authorize" command message.
+   */
+  env = g_getenv ("USER");
+  user = cockpit_hex_encode (env ? env : "", -1);
+  challenge = g_strdup_printf ("plain1:%s:", user);
+  cookie = g_strdup_printf ("askpass%u", (unsigned int)getpid ());
+
+  request = cockpit_transport_build_json ("command", "authorize",
+                                          "challenge", challenge,
+                                          "cookie", cookie,
+                                          NULL);
+
+  if (write_control_message (0, request))
+    {
+      reply = read_control_message (0);
+      if (reply)
+        {
+          if (cockpit_json_get_string (reply, "command", "", &command) &&
+              cockpit_json_get_string (reply, "cookie", "", &field) &&
+              cockpit_json_get_string (reply, "response", "", &response))
```

Comment:
Since this and also the entire `reply` contain the password, it would be nice to `mlock ()` these to avoid it getting swapped out. Of course this would only be race free *before* `cockpit_json_get_string ()` actually writes into it, which the API doesn't allow. To keep things simple and as this is only a short-lived and small program, perhaps just add an `mlockall (MCL_CURRENT | MCL_FUTURE);` at the top of main.